{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### This code takes normalized and binned ChIP-seq data and runs it through Cyber-T statistical analysis.\n",
    "#### The normal model of Cyber-T is not appropriate for count data, but coding this pipeline was proof-of-concept for a later project that I worked on. It also gave me practice in implementing a statistical modeling method as used in Cyber-T. The \"binning\" of ChIP-seq data refers to binning counts across standardized genomic windows with the intent of making differential \"expression\" analysis tractable. The implementation of such code is the subject of an unreleased publication."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Anaconda2\\lib\\site-packages\\pandas\\computation\\__init__.py:19: UserWarning: The installed version of numexpr 2.4.4 is not supported in pandas and will be not be used\n",
      "\n",
      "  UserWarning)\n"
     ]
    }
   ],
   "source": [
    "#=============================================================\n",
    "# 1. Load Data\n",
    "#=============================================================\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# Our data is contained in Norm_Reads\n",
    "Norm_Reads = pd.read_csv(\"QNorm_Reads.txt\",sep=\",\")\n",
    "Norm_Reads = Norm_Reads.set_index(\"Window\")\n",
    "\n",
    "# We load raw reads so we have a reference for ordering the data frames\n",
    "RawReads = pd.read_csv(\"RawAreas.txt\",sep=\"\\t\")\n",
    "RawReads = RawReads.set_index(\"Window\")\n",
    "\n",
    "#=============================================================\n",
    "# 2. Statistical Analysis of Data\n",
    "#=============================================================\n",
    "\n",
    "# Cyber-T implementation\n",
    "# 1. t-statistic: (mu1 - mu2) / SD_bayes\n",
    "# 2. SD_Bayes <- sqrt((conf * SD_prior^2 + (N - 1) * SD_actual^2)/(conf + N - 2)), \n",
    "# where conf is a weight reflecting confidence in data, while SD_actual is the actual SD for the data line\n",
    "# 3. SD_prior = Inferred SD value from a Lowess curve, given a mean\n",
    "# The lowess curve is defined by the curve of best fit for a Mean vs SD plot, for all individual Mean vs SD values \n",
    "\n",
    "Bayes_Weighting = 10\n",
    "\n",
    "import numpy as np\n",
    "from statsmodels.nonparametric.smoothers_lowess import lowess\n",
    "import scipy.stats as stats\n",
    "\n",
    "Norm_Reads_PS = Norm_Reads.loc[:,[\"Norm_PS1\",\"Norm_PS2\"]]\n",
    "Norm_Reads_PS[\"PS_Mean\"] = Norm_Reads_PS[[\"Norm_PS1\",\"Norm_PS2\"]].sum(axis=1)/2.0\n",
    "Norm_Reads_PS[\"PS_SD\"] = np.sqrt((Norm_Reads_PS[\"Norm_PS1\"] - Norm_Reads_PS[\"PS_Mean\"])**2 +  \\\n",
    "                                 (Norm_Reads_PS[\"Norm_PS2\"] - Norm_Reads_PS[\"PS_Mean\"])**2)\n",
    "\n",
    "Norm_Reads_OS = Norm_Reads.loc[:,[\"Norm_OS1\",\"Norm_OS2\"]]\n",
    "Norm_Reads_OS[\"OS_Mean\"] = Norm_Reads_OS[[\"Norm_OS1\",\"Norm_OS2\"]].sum(axis=1)/2.0\n",
    "Norm_Reads_OS[\"OS_SD\"] = np.sqrt((Norm_Reads_OS[\"Norm_OS1\"] - Norm_Reads_OS[\"OS_Mean\"])**2 +  \\\n",
    "                                 (Norm_Reads_OS[\"Norm_OS2\"] - Norm_Reads_OS[\"OS_Mean\"])**2)\n",
    "\n",
    "# http://statsmodels.sourceforge.net/devel/generated/statsmodels.nonparametric.smoothers_lowess.lowess.html\n",
    "\n",
    "Norm_Reads_PS[\"PS_SD_Prior\"] = lowess(endog = np.array(Norm_Reads_PS[\"PS_SD\"]),exog = np.array(Norm_Reads_PS[\"PS_Mean\"]),\n",
    "                                      delta = 0.01 * (max(np.array(Norm_Reads_PS[\"PS_Mean\"])) - min(np.array(Norm_Reads_PS[\"PS_Mean\"]))),\n",
    "                                      return_sorted = False)\n",
    "Norm_Reads_PS[\"PS_SD_Bayes\"] = np.sqrt((Bayes_Weighting*Norm_Reads_PS[\"PS_SD_Prior\"]**2  +  (2 - 1)*Norm_Reads_PS[\"PS_SD\"])\n",
    "                                       /(Bayes_Weighting + 2 - 2))\n",
    "\n",
    "\n",
    "Norm_Reads_OS[\"OS_SD_Prior\"] = lowess(endog = np.array(Norm_Reads_OS[\"OS_SD\"]),exog = np.array(Norm_Reads_OS[\"OS_Mean\"]),\n",
    "                                      delta = 0.01 * (max(np.array(Norm_Reads_OS[\"OS_Mean\"])) - min(np.array(Norm_Reads_OS[\"OS_Mean\"]))),\n",
    "                                      return_sorted = False)\n",
    "Norm_Reads_OS[\"OS_SD_Bayes\"] = np.sqrt((Bayes_Weighting*Norm_Reads_OS[\"OS_SD_Prior\"]**2  +  (2 - 1)*Norm_Reads_OS[\"OS_SD\"]**2)\n",
    "                                       /(Bayes_Weighting + 2 - 2))\n",
    "\n",
    "\n",
    "Norm_Reads_PostStat = Norm_Reads_PS.join(Norm_Reads_OS)\n",
    "\n",
    "# http://www.statsdirect.com/help/parametric_methods/utt.htm\n",
    "# 2-sided t-test with UNEQUAL VARIANCES\n",
    "Norm_Reads_PostStat[\"DF\"] = ((Norm_Reads_PostStat[\"PS_SD_Bayes\"]**2/2 + Norm_Reads_PostStat[\"OS_SD_Bayes\"]**2/2)**2 \\\n",
    "                          / ((Norm_Reads_PostStat[\"PS_SD_Bayes\"]**2/2)**2 + (Norm_Reads_PostStat[\"OS_SD_Bayes\"]**2/2)**2))\n",
    "Norm_Reads_PostStat[\"D_stat\"] = ((Norm_Reads_PostStat[\"PS_Mean\"] - Norm_Reads_PostStat[\"OS_Mean\"])\n",
    "                                 /np.sqrt(Norm_Reads_PostStat[\"PS_SD_Bayes\"]**2/2 + Norm_Reads_PostStat[\"OS_SD_Bayes\"]**2/2))\n",
    "Norm_Reads_PostStat[\"pval\"] = stats.t.sf(np.abs(Norm_Reads_PostStat[\"D_stat\"]), Norm_Reads_PostStat[\"DF\"])*2\n",
    "\n",
    "#=============================================================\n",
    "# 3. Sort/Organize Data\n",
    "#=============================================================\n",
    "\n",
    "# Split normalized reads by chromosome\n",
    "Human_Chromosomes_2 = [\"chrM\",\"chr1'\",\"chr2'\",\"chr3\",\"chr4\",\"chr5\",\"chr6\",\"chr7\",\"chr8\",\"chr9\",\n",
    "                    \"chr10\",\"chr11\",\"chr12\",\"chr13\",\"chr14\",\"chr15\",\"chr16\",\"chr17\",\n",
    "                    \"chr18\",\"chr19\",\"chr20\",\"chr21\",\"chr22\",\"chrX\",\"chrY\"]\n",
    "Areas_By_Chromosome = {}\n",
    "Areas_By_Chromosome_Raw = {}\n",
    "\n",
    "for i in Human_Chromosomes_2:\n",
    "    Areas_By_Chromosome[i] = Norm_Reads_PostStat.ix[[j for j in Norm_Reads_PostStat.index if i in j]]\n",
    "    Areas_By_Chromosome_Raw[i] = RawReads.ix[[j for j in RawReads.index if i in j]]\n",
    "    Areas_By_Chromosome[i] = Areas_By_Chromosome[i].reindex(list(Areas_By_Chromosome_Raw[i].index.values))\n",
    "    \n",
    "Areas_By_Chromosome[\"chr1\"] = Areas_By_Chromosome[\"chr1'\"]\n",
    "Areas_By_Chromosome[\"chr2\"] = Areas_By_Chromosome[\"chr2'\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#=============================================================\n",
    "# Organize Human Gene Annotations\n",
    "#=============================================================\n",
    "\n",
    "# hg19 annotations, downloaded 4/3/2016\n",
    "# Gene annotations were previously consolidated in such a way that \n",
    "# 1. 5 kB regions upstream and downstream were included in coordinates, and \n",
    "# 2. overlapping genes form combined single entries\n",
    "\n",
    "GeneRegions = open(\"GeneList_Consolidated.csv\",\"r\")\n",
    "num_lines = sum(1 for line in GeneRegions)\n",
    "GeneRegions.close()\n",
    "GeneRegions = open(\"GeneList_Consolidated.csv\",\"r\")\n",
    "Human_Chromosomes = [\"chrM\",\"chr1\",\"chr2\",\"chr3\",\"chr4\",\"chr5\",\"chr6\",\"chr7\",\"chr8\",\"chr9\",\n",
    "                    \"chr10\",\"chr11\",\"chr12\",\"chr13\",\"chr14\",\"chr15\",\"chr16\",\"chr17\",\n",
    "                    \"chr18\",\"chr19\",\"chr20\",\"chr21\",\"chr22\",\"chrX\",\"chrY\"]\n",
    "\n",
    "# First row is a header row - get rid of it \n",
    "Region = GeneRegions.readline()\n",
    "\n",
    "# Create a mapping between gene locations and the normalized chipseq data\n",
    "# Regions_Of_Interest is a dict that contains the following information\n",
    "# key: Entrez IDs of the region of interest\n",
    "# value: a list of tuples (chromosome, window_begin, window_end) that correspond to the row labels \n",
    "# in the normalized chipseq data\n",
    "Regions_Of_Interest = {}\n",
    "for Lines in range(0,num_lines):\n",
    "    Region = GeneRegions.readline().strip()\n",
    "    Region = Region.split(\",\")\n",
    "    if len(Region) < 2:\n",
    "        break\n",
    "    Region_Chromosome = Region[1]\n",
    "    Region_Begin = int(Region[2])\n",
    "    Region_End = int(Region[3])\n",
    "    Region_Label = Region[6]\n",
    "    if Region_Chromosome in Human_Chromosomes:\n",
    "        Chrom_Regions = Areas_By_Chromosome[Region_Chromosome].index\n",
    "    else:\n",
    "        continue\n",
    "    Region_Location = [i for i in Chrom_Regions \n",
    "                       if (int(i.split(\", \")[1]) > Region_Begin and int(i.split(\", \")[2].split(\")\")[0]) < Region_End) or\n",
    "                      (int(i.split(\", \")[2].split(\")\")[0]) > Region_Begin and int(i.split(\", \")[2].split(\")\")[0]) < Region_End) \n",
    "                       or (int(i.split(\", \")[1]) > Region_Begin and int(i.split(\", \")[1]) < Region_End)]\n",
    "    Regions_Of_Interest[Region_Label] = Region_Location\n",
    "    \n",
    "GeneRegions.close()\n",
    "\n",
    "# Backup of the mapping\n",
    "# ONLY ACTIVATE WHEN YOU'RE READY TO BACK SOMETHING UP\n",
    "#import pickle\n",
    "\n",
    "#pickle.dump(Regions_Of_Interest,open(\"RegionMapping_Backup.p\",\"wb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#=============================================================\n",
    "# Organize Human Gene Annotations, cont.\n",
    "#=============================================================\n",
    "\n",
    "# Reload mapping\n",
    "import pickle\n",
    "\n",
    "Regions_Of_Interest = pickle.load(open(\"RegionMapping_Backup.p\",\"rb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#=============================================================\n",
    "# DATA VISUALIZATION\n",
    "#=============================================================\n",
    "\n",
    "# Pick individual genes to visualize, by Entrez ID\n",
    "# Entrez ID of KLF2 is 10365, for example. Entrez ID of KLF4 is 9314.\n",
    "# Make a plot of selected data of interest (which also shows gene annotation superimposed over data)\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "\n",
    "Desired_IDs = [3691]\n",
    "\n",
    "for i in Desired_IDs:\n",
    "    Search_For_ID = [elem for elem in Regions_Of_Interest.keys() if (str(i) == elem or str(i)+\":\" in elem or \":\"+str(i) in elem)]\n",
    "    if len(Search_For_ID) == 0:\n",
    "        print i,\" not found\"\n",
    "        continue\n",
    "    Gene_ID = Search_For_ID[0]\n",
    "    \n",
    "    Data_Of_Interest = Norm_Reads_PostStat.ix[Regions_Of_Interest[Gene_ID]]\n",
    "    # 1. Extract chromosomal range and window size\n",
    "    Chr_Range = list(Data_Of_Interest.index.values)\n",
    "    Chromosome = eval(Chr_Range[0])[0]\n",
    "    Window_Size = eval(Chr_Range[0])[2] - eval(Chr_Range[0])[1]\n",
    "    Begin = eval(Chr_Range[0])[1]\n",
    "    End = eval(Chr_Range[-1])[2]\n",
    "\n",
    "    fig = plt.gca()\n",
    "    plt.plot((Begin,End),(0,0),color=\"black\",linewidth=1)\n",
    "    \n",
    "    if \":\" in Gene_ID:\n",
    "        Gene_ID = i\n",
    "\n",
    "    # Plot data\n",
    "    for i in list(Data_Of_Interest.index.values):\n",
    "        Coords = eval(i)\n",
    "        plt.plot((Coords[1], Coords[2]), (Data_Of_Interest.loc[i,\"Norm_OS1\"], Data_Of_Interest.loc[i,\"Norm_OS1\"]),\n",
    "                 color = \"#e74c3c\",linewidth=4)\n",
    "        plt.plot((Coords[1], Coords[2]), (Data_Of_Interest.loc[i,\"Norm_OS2\"], Data_Of_Interest.loc[i,\"Norm_OS2\"]),\n",
    "                 color = '#7b241c',linewidth=4)\n",
    "        plt.plot((Coords[1], Coords[2]), (Data_Of_Interest.loc[i,\"Norm_PS1\"], Data_Of_Interest.loc[i,\"Norm_PS1\"]),\n",
    "                 color = '#3498db',linewidth=4)\n",
    "        plt.plot((Coords[1], Coords[2]), (Data_Of_Interest.loc[i,\"Norm_PS2\"], Data_Of_Interest.loc[i,\"Norm_PS2\"]),\n",
    "                 color = 'blue',linewidth=4)\n",
    "        if float(Data_Of_Interest.loc[i,\"pval\"]) < 0.05:\n",
    "            plt.text((Coords[1]+Coords[2])/2.0-50, max(Data_Of_Interest.loc[i,\"Norm_OS1\"],Data_Of_Interest.loc[i,\"Norm_OS2\"],\n",
    "                                                       Data_Of_Interest.loc[i,\"Norm_PS1\"],Data_Of_Interest.loc[i,\"Norm_PS2\"])+10,\"*\",size=30)\n",
    "\n",
    "    # Plot gene (green portion shows where the TSS is)\n",
    "    Gene_Annotations = pd.read_csv(\"GeneIndex_Ordered.csv\",sep=\",\",index_col=0)\n",
    "    Plotted_Gene_Begin = int(Gene_Annotations.loc[Gene_Annotations[\"gene_id\"] == int(Gene_ID),\"start\"])\n",
    "    Plotted_Gene_End = int(Gene_Annotations.loc[Gene_Annotations[\"gene_id\"] == int(Gene_ID),\"end\"])\n",
    "    Plotted_Gene_Name = list(Gene_Annotations.loc[Gene_Annotations[\"gene_id\"] == int(Gene_ID),\"gene_name\"])[0]\n",
    "    plt.plot((Plotted_Gene_Begin+5000, Plotted_Gene_End-5000), (-1000, -1000),color = 'black',linewidth=6)\n",
    "\n",
    "    if (Gene_Annotations.loc[Gene_Annotations[\"gene_id\"] == int(Gene_ID),\"strand\"] == \"+\").any():\n",
    "        plt.plot((Plotted_Gene_Begin+5000, Plotted_Gene_Begin+5010), (-1000, -1000),color = 'green',linewidth=6)\n",
    "        plt.text(Plotted_Gene_Begin+5000,-750,Plotted_Gene_Name)\n",
    "    else:\n",
    "        plt.plot((Plotted_Gene_End-5010, Plotted_Gene_End-5000), (-1000, -1000),color = 'green',linewidth=6)\n",
    "        plt.text(Plotted_Gene_Begin+5000,-1250,Plotted_Gene_Name)\n",
    "\n",
    "    # Labels and formatting\n",
    "    fig.get_xaxis().get_major_formatter().set_useOffset(False)\n",
    "    fig.get_xaxis().get_major_formatter().set_scientific(False)\n",
    "    fig.set_xlabel(Chromosome,size=18)\n",
    "    fig.set_ylabel(\"Normalized area under curve per 1000-nt window\",size=15)\n",
    "    fig.tick_params(axis='both', which='major', labelsize=15)\n",
    "\n",
    "    # Legend\n",
    "    OS1_patch = mpatches.Patch(color='#e74c3c', label='OS1')\n",
    "    OS2_patch = mpatches.Patch(color='#7b241c', label='OS2')\n",
    "    PS1_patch = mpatches.Patch(color='#3498db', label='PS1')\n",
    "    PS2_patch = mpatches.Patch(color='blue', label='PS2')\n",
    "    TSS_patch = mpatches.Patch(color='green', label='TSS')\n",
    "    plt.legend(handles=[OS1_patch,OS2_patch,PS1_patch,PS2_patch,TSS_patch])\n",
    "\n",
    "    plt.show()\n",
    "        \n",
    "\n",
    "#Regions_Of_Interest[\"10365\"]\n",
    "\n",
    "#Data_Of_Interest = Norm_Reads_PostStat.ix[Regions_Of_Interest[\"10365\"]]\n",
    "\n",
    "#Data_Of_Interest.to_csv(\"KLF2.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'Norm_Reads_PS' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-3-6d647eba5d72>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[1;31m#=============================================================\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 5\u001b[1;33m \u001b[0mNormPS_Sample\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mNorm_Reads_PS\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msample\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mn\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m5000\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      6\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mmatplotlib\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpyplot\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mplt\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      7\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mmatplotlib\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpatches\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mmpatches\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'Norm_Reads_PS' is not defined"
     ]
    }
   ],
   "source": [
    "#=============================================================\n",
    "# DATA VISUALIZATION - CHECKING LOWESS FUNCTION ON DATA\n",
    "#=============================================================\n",
    "\n",
    "NormPS_Sample = Norm_Reads_PS.sample(n=5000)\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "\n",
    "NormPS_Sample.plot.scatter(x=\"PS_Mean\",y=\"PS_SD\",xlim=(0,3500),ylim=(0,3500))\n",
    "NormPS_Sample.plot.scatter(x=\"PS_Mean\",y=\"PS_SD_Prior\",xlim=(0,3500),ylim=(0,3500))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "441058  not found\n",
      "727764  not found\n",
      "7146  not found\n",
      "389831  not found\n",
      "403340  not found\n"
     ]
    }
   ],
   "source": [
    "#=============================================================\n",
    "# DATA FILTRATION\n",
    "#=============================================================\n",
    "\n",
    "# Find all genes with at least one region that is differentially acetylated.\n",
    "Gene_Annotations = pd.read_csv(\"GeneIndex_Ordered.csv\",sep=\",\",index_col=0)\n",
    "All_IDs = list(Gene_Annotations[\"gene_id\"])\n",
    "\n",
    "Good_IDs = []\n",
    "for i in All_IDs:\n",
    "    Search_For_ID = [elem for elem in Regions_Of_Interest.keys() if (str(i) == elem or str(i)+\":\" in elem or \":\"+str(i) in elem)]\n",
    "    if len(Search_For_ID) == 0:\n",
    "        print i,\" not found\"\n",
    "        continue\n",
    "    Gene_ID = Search_For_ID[0]\n",
    "    \n",
    "    Data_Of_Interest = Norm_Reads_PostStat.ix[Regions_Of_Interest[Gene_ID]]\n",
    "    \n",
    "    if Data_Of_Interest.loc[Data_Of_Interest[\"pval\"] < 0.05,\"pval\"].count() > 0:\n",
    "        Good_IDs += [i]\n",
    "    \n",
    "Good_Gene_Information = Gene_Annotations.ix[Good_IDs]\n",
    "\n",
    "Good_Gene_Information.to_csv(\"GenesWithDEAcetylation.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from statsmodels.nonparametric.smoothers_lowess import lowess\n",
    "import scipy.stats as stats\n",
    "from copy import deepcopy\n",
    "import ast\n",
    "import math\n",
    "\n",
    "# Generalized version of statistical analysis - debugged\n",
    "\n",
    "\n",
    "def StatAnalysis_OfQNormBins(QNormArea_File, RawArea_File, QNormReadsWithPVals_File, Control_Sets, Experimental_Sets,\n",
    "                             annotation_file, Bayes_Weighting=10):\n",
    "    # =============================================================\n",
    "    # 1. Load Data\n",
    "    # ============================================================\n",
    "\n",
    "    # Our data is contained in Norm_Reads\n",
    "    Norm_Reads = pd.read_csv(QNormArea_File, sep=\",\")\n",
    "    Norm_Reads = Norm_Reads.set_index(\"Window\")\n",
    "\n",
    "    # We load raw reads so we have a reference for ordering the data frames\n",
    "    # Note 6/13/17: Norm_Reads are reordered in the normalization function, so this step is not necessary\n",
    "    #RawReads = pd.read_csv(RawArea_File, sep=\"\\t\")\n",
    "    #RawReads = RawReads.set_index(\"Window\")\n",
    "\n",
    "    # =============================================================\n",
    "    # 2. Statistical Analysis of Data\n",
    "    # =============================================================\n",
    "\n",
    "    # Cyber-T implementation\n",
    "    # According to Shakti:\n",
    "    # 1. t-statistic: (mu1 - mu2) / SD_bayes\n",
    "    # 2. SD_Bayes <- sqrt((conf * SD_prior^2 + (N - 1) * SD_actual^2)/(conf + N - 2)),\n",
    "    # where conf is a weight reflecting confidence in data, while SD_actual is the actual SD for the data line\n",
    "    # 3. SD_prior = Inferred SD value from a Lowess curve, given a mean\n",
    "    # The lowess curve is defined by the curve of best fit for a Mean vs SD plot, for all individual Mean vs SD values\n",
    "\n",
    "    # Bayes_Weighting = 10\n",
    "\n",
    "    Norm_Reads_Control = Norm_Reads.loc[:, Control_Sets]\n",
    "    Norm_Reads_Control[\"Ctrl_Mean\"] = Norm_Reads_Control[Control_Sets].sum(axis=1) / float(len(Control_Sets))\n",
    "\n",
    "    temp = deepcopy(Norm_Reads_Control)\n",
    "    for column in Control_Sets:\n",
    "        temp[column] = temp[column] - temp[\"Ctrl_Mean\"]\n",
    "    Norm_Reads_Control[\"Ctrl_SD\"] = np.sqrt((temp[Control_Sets] ** 2).sum(axis=1))\n",
    "\n",
    "    Norm_Reads_Experiment = Norm_Reads.loc[:, Experimental_Sets]\n",
    "    Norm_Reads_Experiment[\"Expt_Mean\"] = Norm_Reads_Experiment[Experimental_Sets].sum(axis=1) / float(len(Experimental_Sets))\n",
    "    temp2 = deepcopy(Norm_Reads_Experiment)\n",
    "    for column in Experimental_Sets:\n",
    "        temp2[column] = temp2[column] - temp2[\"Expt_Mean\"]\n",
    "    Norm_Reads_Experiment[\"Expt_SD\"] = np.sqrt((temp2[Experimental_Sets] ** 2).sum(axis=1))\n",
    "\n",
    "    # http://statsmodels.sourceforge.net/devel/generated/statsmodels.nonparametric.smoothers_lowess.lowess.html\n",
    "\n",
    "    Norm_Reads_Control[\"Ctrl_SD_Prior\"] = lowess(endog=np.array(Norm_Reads_Control[\"Ctrl_SD\"]),\n",
    "                                                 exog=np.array(Norm_Reads_Control[\"Ctrl_Mean\"]),\n",
    "                                                 delta=0.01 * (max(np.array(Norm_Reads_Control[\"Ctrl_Mean\"])) - min(\n",
    "                                                     np.array(Norm_Reads_Control[\"Ctrl_Mean\"]))),\n",
    "                                                 return_sorted=False)\n",
    "    Norm_Reads_Control[\"Ctrl_SD_Bayes\"] = np.sqrt(\n",
    "        (Bayes_Weighting * Norm_Reads_Control[\"Ctrl_SD_Prior\"] ** 2 + (len(Control_Sets) - 1) * Norm_Reads_Control[\"Ctrl_SD\"] ** 2)\n",
    "        / (Bayes_Weighting + len(Control_Sets) - 2.0))\n",
    "\n",
    "    Norm_Reads_Experiment[\"Expt_SD_Prior\"] = lowess(endog=np.array(Norm_Reads_Experiment[\"Expt_SD\"]),\n",
    "                                                    exog=np.array(Norm_Reads_Experiment[\"Expt_Mean\"]),\n",
    "                                                    delta=0.01 * (\n",
    "                                                    max(np.array(Norm_Reads_Experiment[\"Expt_Mean\"])) - min(\n",
    "                                                        np.array(Norm_Reads_Experiment[\"Expt_Mean\"]))),\n",
    "                                                    return_sorted=False)\n",
    "    Norm_Reads_Experiment[\"Expt_SD_Bayes\"] = np.sqrt(\n",
    "        (Bayes_Weighting * Norm_Reads_Experiment[\"Expt_SD_Prior\"] ** 2 + (len(Experimental_Sets) - 1) *\n",
    "         Norm_Reads_Experiment[\"Expt_SD\"] ** 2)\n",
    "        / (Bayes_Weighting + len(Experimental_Sets) - 2.0))\n",
    "\n",
    "    Norm_Reads_PostStat = Norm_Reads_Control.join(Norm_Reads_Experiment)\n",
    "\n",
    "    # http://www.statsdirect.com/help/parametric_methods/utt.htm\n",
    "    # 2-sided t-test with UNEQUAL VARIANCES\n",
    "    Norm_Reads_PostStat[\"DF\"] = (\n",
    "        ((Norm_Reads_PostStat[\"Ctrl_SD_Bayes\"] ** 2 / float(len(Control_Sets))\n",
    "          + Norm_Reads_PostStat[\"Expt_SD_Bayes\"] ** 2 / float(len(Experimental_Sets))) ** 2)\n",
    "        / (\n",
    "            (((Norm_Reads_PostStat[\"Ctrl_SD_Bayes\"] ** 2) / (float(len(Control_Sets)) ** 2)) / (\n",
    "            float(len(Control_Sets) - 1.0)))\n",
    "            + (((Norm_Reads_PostStat[\"Expt_SD_Bayes\"] ** 2) / (float(len(Experimental_Sets)) ** 2)) / (\n",
    "            float(len(Experimental_Sets) - 1.0)))\n",
    "        )\n",
    "    )\n",
    "\n",
    "    Norm_Reads_PostStat[\"D_stat\"] = ((Norm_Reads_PostStat[\"Ctrl_Mean\"] - Norm_Reads_PostStat[\"Expt_Mean\"])\n",
    "                                     / np.sqrt(Norm_Reads_PostStat[\"Ctrl_SD_Bayes\"] ** 2 / float(len(Control_Sets))\n",
    "                                               + Norm_Reads_PostStat[\"Expt_SD_Bayes\"] ** 2 / float(len(Experimental_Sets)))\n",
    "                                     )\n",
    "\n",
    "    Norm_Reads_PostStat[\"pval\"] = stats.t.sf(np.abs(Norm_Reads_PostStat[\"D_stat\"]), Norm_Reads_PostStat[\"DF\"]) * 2\n",
    "\n",
    "    Norm_Reads_PostStat[\"bonf\"] = Norm_Reads_PostStat[\"pval\"] * Norm_Reads_PostStat.shape[0]\n",
    "    Norm_Reads_PostStat.loc[Norm_Reads_PostStat[\"bonf\"] > 1,\"bonf\"] = 1\n",
    "\n",
    "    # =============================================================\n",
    "    # 3. Exporting And Binning Data\n",
    "    # =============================================================\n",
    "\n",
    "    # The analyzed data file will be saved.\n",
    "    # Data binned by chromosome will also be saved, for quick plotting access.\n",
    "\n",
    "    Norm_Reads_PostStat.to_csv(QNormReadsWithPVals_File)\n",
    "\n",
    "    Columns_Of_Importance = Norm_Reads.columns.tolist() + [\"pval\",\"bonf\"]\n",
    "    Condensed_NormReads = Norm_Reads_PostStat[Columns_Of_Importance]\n",
    "    condensed_filename = QNormReadsWithPVals_File.split(\".txt\")[0] + \"_condensed.csv\"\n",
    "\n",
    "    Chromosomes = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,\"X\",\"Y\"]\n",
    "\n",
    "    for i in Chromosomes:\n",
    "        StringSearch = \"('chr\" + str(i) + \"',\"\n",
    "        Chrom_Windows = [j for j in list(Condensed_NormReads.index) if StringSearch in j]\n",
    "        if len(Chrom_Windows) > 0:\n",
    "            Chromosome_Data = Condensed_NormReads.ix[Chrom_Windows]\n",
    "            chrom_filename = QNormReadsWithPVals_File.split(\".txt\")[0] + \"_chr\" + str(i) + \"_condensed.csv\"\n",
    "            Chromosome_Data.to_csv(chrom_filename)\n",
    "\n",
    "    Window = ast.literal_eval(Chrom_Windows[0])\n",
    "    WindowSize = Window[2] - Window[1]\n",
    "\n",
    "\n",
    "    Condensed_NormReads.to_csv(condensed_filename)\n",
    "    \n",
    "    return Condensed_NormReads, WindowSize\n",
    "    \n",
    "def Testing_GeneFile_Generation(annotation_file, Condensed_NormReads, Control_Sets, Experimental_Sets,QNormReadsWithPVals_File,\n",
    "                               WindowSize):\n",
    "\n",
    "    # =============================================================\n",
    "    # 4. Finding all differentially modified gene regions\n",
    "    # =============================================================\n",
    "\n",
    "    # Possible ways to summarize a gene, given multi-window data:\n",
    "    # (1) Min p-value, with corresponding LFC\n",
    "    # (2) Max LFC, with corresponding p-value\n",
    "    # Only (1) will be implemented for now.\n",
    "\n",
    "    Annotation_File = annotation_file.set_index(\"Gene stable ID\")\n",
    "    List_Of_Annotated_Genes = list(Annotation_File.index)\n",
    "    List_Of_Annotated_Genes = [x.encode('UTF8') for x in List_Of_Annotated_Genes]\n",
    "\n",
    "    GeneData = []\n",
    "\n",
    "    for Ensembl_ID in List_Of_Annotated_Genes:\n",
    "        Chromosome = Annotation_File.loc[Ensembl_ID, \"Chromosome/scaffold name\"]\n",
    "        GeneBegin = Annotation_File.loc[Ensembl_ID, \"Gene start (bp)\"]\n",
    "        GeneEnd = Annotation_File.loc[Ensembl_ID, \"Gene end (bp)\"]\n",
    "        Strand = Annotation_File.loc[Ensembl_ID, \"Strand\"]\n",
    "        Name = Annotation_File.loc[Ensembl_ID, \"Gene name\"]\n",
    "        \n",
    "        Chromosomes = [\"1\",\"2\",\"3\",\"4\",\"5\",\"6\",\"7\",\"8\",\"9\",\"10\",\"11\",\"12\",\n",
    "                       \"13\",\"14\",\"15\",\"16\",\"17\",\"18\",\"19\",\"20\",\"21\",\"22\",\"X\",\"Y\"]\n",
    "        \n",
    "        if str(Chromosome) not in Chromosomes:\n",
    "            continue\n",
    "        else:\n",
    "            GeneWindowBegin = GeneBegin - 5000\n",
    "            GeneWindowEnd = GeneEnd + 5000\n",
    "\n",
    "            # From gene annotation data, define the regions of interest in (chromosome, begin, end) format\n",
    "            DataFromGeneWindowBegin = int(math.floor(GeneWindowBegin / float(WindowSize)) * float(WindowSize))\n",
    "            DataFromGeneWindowEnd = int(math.ceil(GeneWindowEnd / float(WindowSize)) * float(WindowSize))\n",
    "            Regions_Of_Interest = []\n",
    "            i = DataFromGeneWindowBegin\n",
    "            while i < DataFromGeneWindowEnd:\n",
    "                Region = \"('chr\" + str(Chromosome) + \"', \" + str(i) + \", \" + str(i + WindowSize) + \")\"\n",
    "                Regions_Of_Interest += [Region]\n",
    "                i += WindowSize\n",
    "\n",
    "            Data_Of_Interest = Condensed_NormReads.ix[Regions_Of_Interest]\n",
    "            Min_PVal = Data_Of_Interest[\"bonf\"].min()\n",
    "            #print Min_PVal\n",
    "            #print Annotation_File.loc[Ensembl_ID,:]\n",
    "        \n",
    "            Control_Sum = Data_Of_Interest.loc[Data_Of_Interest[\"bonf\"] == Min_PVal,:]\n",
    "        \n",
    "            # Account for case where gene has multiple min p-vals of same value \n",
    "            # (usual case: gene is not DE and has p values of 1 across all windows)\n",
    "            # The fold change to be reported for this case will be the fold change of largest magnitude\n",
    "            if isinstance(Control_Sum, pd.DataFrame):\n",
    "                Control_Sum = Data_Of_Interest.loc[Data_Of_Interest[\"bonf\"] == Min_PVal,Control_Sets].sum(axis=1) + 1.0\n",
    "                Expt_Sum = Data_Of_Interest.loc[Data_Of_Interest[\"bonf\"] == Min_PVal,Experimental_Sets].sum(axis=1) + 1.0\n",
    "                LogFoldChanges = np.log2(Expt_Sum/Control_Sum)\n",
    "                LargestFoldChangeLocation = LogFoldChanges.abs().idxmax()\n",
    "                LogFoldChange = LogFoldChanges[LargestFoldChangeLocation]\n",
    "            else:\n",
    "                Control_Sum = sum(Data_Of_Interest.loc[Data_Of_Interest[\"bonf\"] == Min_PVal,Control_Sets].tolist()) + 1.0\n",
    "                Expt_Sum = sum(Data_Of_Interest.loc[Data_Of_Interest[\"bonf\"] == Min_PVal,Experimental_Sets].tolist()) + 1.0\n",
    "                FoldChange = Control_Sum / Expt_Sum\n",
    "                LogFoldChange = np.log2(FoldChange)\n",
    "\n",
    "            DataRow = [Ensembl_ID, Name,Chromosome,LogFoldChange,Min_PVal]\n",
    "            GeneData += [DataRow]\n",
    "\n",
    "    Final_GeneData = pd.DataFrame(GeneData,\n",
    "                                  columns=[\"Ensembl ID\",\"Name\",\"Chromosome\",\"Most significant log2FC\",\"Adj. P Value\"])\n",
    "\n",
    "    genedata_filename = QNormReadsWithPVals_File.split(\".txt\")[0] + \"_genesummary.csv\"\n",
    "    Final_GeneData.to_csv(genedata_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>HGNC symbol</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Gene name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14412</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>14363</td>\n",
       "      <td>29806</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>MIR1302-10</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>MIR1302-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>FAM138A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000268020</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>52473</td>\n",
       "      <td>54936</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>OR4G4P</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene stable ID HGNC symbol  Gene start (bp)  Gene end (bp)  Strand  \\\n",
       "0  ENSG00000223972     DDX11L1            11869          14412       1   \n",
       "1  ENSG00000227232      WASH7P            14363          29806      -1   \n",
       "2  ENSG00000243485  MIR1302-10            29554          31109       1   \n",
       "3  ENSG00000237613     FAM138A            34554          36081      -1   \n",
       "4  ENSG00000268020      OR4G4P            52473          54936       1   \n",
       "\n",
       "  Chromosome/scaffold name   Gene name  \n",
       "0                        1     DDX11L1  \n",
       "1                        1      WASH7P  \n",
       "2                        1  MIR1302-10  \n",
       "3                        1     FAM138A  \n",
       "4                        1      OR4G4P  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Annotation_File = pd.read_csv(\"Human GRCh37.p13 07.19.17.csv\")\n",
    "#Annotation_File = Annotation_File.set_index(\"Gene stable ID\")\n",
    "Annotation_File.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>D_stat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Window</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>('chr19', 52903000, 52904000)</th>\n",
       "      <td>5.339418</td>\n",
       "      <td>0.091882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>('chr8', 36265000, 36266000)</th>\n",
       "      <td>0.411270</td>\n",
       "      <td>0.722079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>('chr7', 25503000, 25504000)</th>\n",
       "      <td>-4.208745</td>\n",
       "      <td>0.085011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>('chr3', 20704000, 20705000)</th>\n",
       "      <td>3.377282</td>\n",
       "      <td>0.113791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>('chr6', 165394000, 165395000)</th>\n",
       "      <td>-2.029002</td>\n",
       "      <td>0.189813</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  D_stat      pval\n",
       "Window                                            \n",
       "('chr19', 52903000, 52904000)   5.339418  0.091882\n",
       "('chr8', 36265000, 36266000)    0.411270  0.722079\n",
       "('chr7', 25503000, 25504000)   -4.208745  0.085011\n",
       "('chr3', 20704000, 20705000)    3.377282  0.113791\n",
       "('chr6', 165394000, 165395000) -2.029002  0.189813"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Norm_Reads_PostStat[[\"D_stat\",\"pval\"]].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>D_stat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Window</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>('chrM', 0, 1000)</th>\n",
       "      <td>4.141372</td>\n",
       "      <td>0.000035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>('chrM', 1000, 2000)</th>\n",
       "      <td>3.854526</td>\n",
       "      <td>0.000116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>('chrM', 2000, 3000)</th>\n",
       "      <td>3.936609</td>\n",
       "      <td>0.000083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>('chrM', 3000, 4000)</th>\n",
       "      <td>4.746932</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>('chrM', 4000, 5000)</th>\n",
       "      <td>4.077633</td>\n",
       "      <td>0.000045</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        D_stat      pval\n",
       "Window                                  \n",
       "('chrM', 0, 1000)     4.141372  0.000035\n",
       "('chrM', 1000, 2000)  3.854526  0.000116\n",
       "('chrM', 2000, 3000)  3.936609  0.000083\n",
       "('chrM', 3000, 4000)  4.746932  0.000002\n",
       "('chrM', 4000, 5000)  4.077633  0.000045"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Test_General_Code.loc[RawReads.index,[\"D_stat\",\"pval\"]].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
